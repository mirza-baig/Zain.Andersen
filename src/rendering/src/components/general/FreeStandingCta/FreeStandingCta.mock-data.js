const defaultData = {
    "rendering": {
        "componentName": "FreeStandingCta",
        "dataSource": "/sitecore/content/AndersenCorporation/AndersenWindows/AndersenWindows/Home/Generic Page Mashup1/Data/New Free Standing CTA"
    },
    "fields": {
        "alignment": null,
        "anchorName": {
            "value": ""
        },
        "gaAction": {
            "value": ""
        },
        "gaCategory": {
            "value": ""
        },
        "gaClickedUrlForLabel": {
            "value": false
        },
        "sectionId": {
            "value": ""
        },
        "componentSpacing": null,
        "cta1AriaLabel": {
            "value": ""
        },
        "cta1Icon": {
            "id": "50590edc-7ea7-4436-9a3e-701c87a07db2",
            "url": "https://cm.local.ew.andersencorp.com/sitecore/login/sitecore/system/Settings/Foundation/EnterpriseWeb/Enums/Icons/Arrow",
            "name": "Arrow",
            "displayName": "Arrow",
            "fields": {
                "Value": {
                    "value": "arrow"
                }
            },
            "templateId": "d2923fee-da4e-49be-830c-e27764dfa269",
            "templateName": "Enum"
        },
        "cta1Link": {
            "value": {
                "href": "/",
                "text": "Home",
                "anchor": "",
                "linktype": "internal",
                "class": "",
                "title": "",
                "target": "",
                "querystring": "",
                "id": "{7FB335D2-8E99-458E-9EF9-562A78CCB821}"
            }
        },
        "cta1Modal": null,
        "cta1ModalLinkText": {
            "value": ""
        },
        "cta1Style": {
            "id": "49a23327-0397-4cce-a930-e76918d37c42",
            "url": "https://cm.local.ew.andersencorp.com/sitecore/login/sitecore/system/Settings/Foundation/EnterpriseWeb/Enums/CTA-Styles/Primary",
            "name": "Primary",
            "displayName": "Primary",
            "fields": {
                "Value": {
                    "value": "primary"
                }
            },
            "templateId": "d2923fee-da4e-49be-830c-e27764dfa269",
            "templateName": "Enum"
        },
        "eventType": {
            "value": ""
        },
        "eventZone": {
            "value": ""
        },
        "eventName": {
            "value": ""
        }
    }
};

export default defaultData;
