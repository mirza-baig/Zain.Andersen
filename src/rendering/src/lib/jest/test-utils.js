// Global
import { render, fireEvent, waitFor } from '@testing-library/react';
import { ComponentPropsContext } from '@sitecore-jss/sitecore-jss-nextjs';
import { FastSitecoreContext } from 'lib/overrides/FastSitecoreContext';

// Lib
import { createSampleComponentFactory } from 'lib/mocks/mock-placeholder';

export const eeSitecoreContext = {
  route: {
    name: 'Foo',
    placeholders: {},
  },
  context: {
    pageEditing: true,
    site: {
      name: 'JssTestWeb',
    },
    language: 'en',
  },
};

export const renderComponent = (
  Component,
  {
    componentProps = {},
    staticProps = {},
    componentFactory = createSampleComponentFactory(),
    sitecoreContext = {
      route: {
        name: 'Foo',
        placeholders: {},
      },
    },
  }
) => {
  const mockLayoutData = {
    sitecore: sitecoreContext,
  };

  return render(
    <ComponentPropsContext value={staticProps}>
      <FastSitecoreContext componentFactory={componentFactory} layoutData={mockLayoutData}>
        <Component {...componentProps} />
      </FastSitecoreContext>
    </ComponentPropsContext>
  );
};

export const snapshot = (Component, mockData) => {
  const component = renderComponent(Component, mockData);
  expect(component.container.firstChild).toMatchSnapshot();
  return component;
};

export const hasDataComponent = async (component, name) => {
  await waitFor(() => {
    const nodes = component.baseElement.querySelectorAll(`[data-component="${name}"]`);
    return expect(nodes.length).toBe(1);
  });
};

export const mockDateAs = (epochMs) => {
  if (typeof epochMs !== 'number') {
    console.error(
      'mockDateAs only accepts epoch values in milliseconds to prevent timezones from factoring in to Date creation.'
    );
  }

  jest.spyOn(global.Date, 'now').mockImplementation(() => new Date(epochMs).valueOf());
};

export const testFieldError = async (Component, mockData, { label, value, error }) => {
  const component = renderComponent(Component, mockData);
  const field = component.getByLabelText(label);
  fireEvent.change(field, { target: { value } });
  field.blur();
  await waitFor(() => {
    const err = component.queryByText(error);
    return expect(err).not.toBe(null);
  });
};
