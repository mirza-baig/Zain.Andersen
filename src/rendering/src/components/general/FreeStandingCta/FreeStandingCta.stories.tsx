// Global
import { Story, Meta } from '@storybook/react';
// Local
import FreeStandingCta, { FreeStandingCtaProps } from './FreeStandingCta';
import defaultData from './FreeStandingCta.mock-data';

export default {
  title: 'General/FreeStandingCta',
  component: FreeStandingCta,
} as Meta;
const Template: Story<FreeStandingCtaProps> = (props) => <FreeStandingCta {...props} />;
export const Default = Template.bind({});
Default.args = defaultData;
