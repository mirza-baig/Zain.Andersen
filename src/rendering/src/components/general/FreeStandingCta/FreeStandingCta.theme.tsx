// Lib
import { ThemeFile } from 'lib/context/ThemeContext';
import { ComponentAlignment } from '../FreeStandingCta/FreeStandingCta';
import classNames from 'classnames';

const getDynamicStyles = (alignment: ComponentAlignment): string => {
  let classes = 'col-span-12';
  if (alignment === 'left') {
    classes = classNames('col-span-12');
  } else if (alignment === 'center') {
    classes = classNames('col-span-12 col-start-2 md:col-span-12 md:col-start-6');
  } else {
    classes = classNames('col-span-12 col-start-12 md:col-span-12 md:col-start-12');
  }
  return classes;
};

export const FreeStandingCtaTheme = (alignment: ComponentAlignment): ThemeFile => {
  return {
    aw: {
      classes: {
        contentWrapper: classNames(getDynamicStyles(alignment)),
        'col-start-12': {
          'grid-column-start': 12,
        },
      },
    },
    rba: {
      classes: {
        /** Insert Theme classes here **/
      },
    },
  };
};
/* Example of more advanced builder */
/*
import classNames from 'classnames';
const getDynamicStyles = (theme: ThemeName): string => {
  return classNames('text-xl', theme === 'aw' ? 'color-orange' : 'color-green');
};
export const FreeStandingCtaTheme = (): ThemeFile => {
  return {
    aw: {
      classes: {
        // Insert Theme classes here  - sample function call
        headline: classNames(getDynamicStyles('aw')),
      },
    },
    rba: {
      classes: {
        // Insert Theme classes here  - sample function call
        headline: classNames(getDynamicStyles('rba')),
      },
    },
  };
};
*/
