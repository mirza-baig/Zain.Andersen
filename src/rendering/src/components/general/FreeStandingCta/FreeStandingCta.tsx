// Global
import { Feature } from 'src/.generated/Feature.EnterpriseWeb.model';
import { withDatasourceCheck, LinkField } from '@sitecore-jss/sitecore-jss-nextjs';
import { useTheme } from 'lib/context/ThemeContext';
// Components
import { Component } from 'src/helpers/Component';
import SingleButton from 'src/helpers/SingleButton/SingleButton';
import { FreeStandingCtaTheme } from './FreeStandingCta.theme';
import { getEnum } from 'lib/utils';
import { hashCode } from 'src/helpers/Component/Component';

export type FreeStandingCtaProps =
  Feature.EnterpriseWeb.AndersenWindows.Components.General.FreeStandingCta.FreeStandingCta;
export type ComponentAlignment = 'left' | 'center';

const FreeStandingCta = (props: FreeStandingCtaProps): JSX.Element => {
  const linkField = props?.fields.cta1Link as LinkField;
  const alignment = getEnum<ComponentAlignment>(props.fields?.alignment) || 'left';
  const { themeData } = useTheme(FreeStandingCtaTheme(alignment));

  return (
    <Component
      variant="lg"
      dataComponent="general/freestandingcta"
      {...props}
      spacing={props.fields?.spacing}
    >
      <section
        className={themeData.classes.contentWrapper}
        id={props.fields?.sectionId?.value || `id${hashCode(props.rendering?.dataSource)}`}
      >
        <div className="mb-s flex items-start md:flex-row md:items-center md:space-x-4">
          {linkField.value.href && (
            <SingleButton {...props} classes={themeData.classes.sectionCta} />
          )}
        </div>
      </section>
    </Component>
  );
};

export default withDatasourceCheck()<FreeStandingCtaProps>(FreeStandingCta);
