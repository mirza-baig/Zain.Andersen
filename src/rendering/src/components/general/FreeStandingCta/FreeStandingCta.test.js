// Global
import { hasDataComponent, snapshot } from 'lib/jest/test-utils';
// Local
import FreeStandingCta from './FreeStandingCta';
import defaultData from './FreeStandingCta.mock-data';

it('renders correctly', async () => {
  const component = snapshot(FreeStandingCta, { componentProps: defaultData });
  await hasDataComponent(component, 'general/freestandingcta');
});
