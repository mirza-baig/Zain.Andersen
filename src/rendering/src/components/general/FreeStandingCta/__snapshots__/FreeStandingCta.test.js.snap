// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  data-component="general/freestandingcta"
  id="id175593572"
>
  <div
    class="my-8 lg:mx-auto md:max-w-screen-lg theme-white"
  >
    <div
      class="section-grid grid grid-cols-2 md:grid-cols-12 px-m gap-s md:gap-s "
    >
      <section
        class="col-span-12"
        id="id175593572"
      >
        <div
          class="mb-s flex items-start md:flex-row md:items-center md:space-x-4"
        >
          <div
            class="mb-s flex items-start md:flex-row md:items-center"
          >
            <a
              aria-label="Home"
              class="flex w-fit items-center whitespace-normal rounded-lg border-4 border-theme-btn-border bg-theme-btn-bg px-m py-[9px] font-sans text-button font-heavy text-theme-btn-text hover:border-theme-btn-border-hover hover:bg-theme-btn-bg-hover hover:text-theme-btn-text-hover disabled:border-gray disabled:text-gray disabled:hover:bg-theme-btn-bg disabled:cursor-not-allowed"
              data-gtm-click=""
              data-gtm-dl-event="cta_click"
              href="/"
              role="link"
              tabindex="0"
              target=""
              title=""
            >
              Home
              <span
                class="ml-xxs"
              />
            </a>
          </div>
        </div>
      </section>
    </div>
  </div>
</section>
`;
